### ISSUE TITLE
Using JSON to describe opcodes

### ISSUE THREAD

**Issue by nwhetsell** [csound/csound#536]
I’m writing a [`language-csound` package](https://atom.io/packages/language-csound) for [Atom](https://atom.io). One thing the package does is autocomplete Csound’s built-in opcodes. In working on this, I found that there doesn’t seem to be a standard, easily-readable way of describing opcodes—what they do, what types of arguments they take, what range of input arguments is valid, what the defaults are, _etc_. Since this may touch on some of the bullets of [Topics at the round table ICSC 2015](https://github.com/csound/csound/issues/533) (specifically “Learning resources” and “Csound packages distribution standard”), I thought I’d ask about this here.

My thought is to describe opcodes using JSON. Here’s an example using the [`oscili`](http://www.csounds.com/manual/html/oscili.html) opcode.

```json
{ "name": "oscili",
  "description": "Oscillator with linear interpolation",
  "descriptionURL": "http://www.csounds.com/manual/html/oscili.html",
  "inputs": [
    { "name": "Amplitude" },
    { "name": "Frequency",
      "units": "Hz"
    },
    { "name": "Function Table ID",
      "min": -1,
      "increment": 1,
      "default": -1
    },
    { "name": "Initial Phase",
      "units": "cycles",
      "min": 0,
      "max": 1,
      "default": 0
    }
  ],
  "outputs": [
    { "name": "Signal" }
  ],
  "syntax": [
    { "inputTypes": [
        ["a", "k"],
        ["a", "k"],
        ["i", "i[]"],
        "i"
      ],
      "outputTypes": [
        "a"
      ]
    },
    { "inputTypes": [
        "k",
        "k",
        ["i", "i[]"],
        "i"
      ],
      "outputTypes": [
        "k"
      ]
    }
  ]
}
```

Here’s a description of this proposed format, with a bit more detail for dealing with opcodes that return a variable number of outputs or take a variable number of inputs.

First, let a _type array_ be a JSON array that contains only:
* strings denoting a single input type, like `"a"` or `"k"`, and/or
* arrays of such strings.

The entire opcode description is a dictionary containing:
* `name` — the opcode’s name as a string.
* `description` — a string that briefly describes the opcode.
* `descriptionURL` — a URL for a file with a more complete description of the opcode.
* `inputs` — an array of dictionaries describing the opcode’s inputs, in order. The dictionaries can contain
  * `name` — a brief, title-case string describing the input, suitable for presentation in a user-interface.
  * `units` — a string giving the units of the input, suitable for presentation in a user-interface.
  * `min` — a number giving the minimum value of a numerical input.
  * `max` — a number giving the maximum value of a numerical input.
  * `increment` — a number specifying the difference between numerical inputs. A minimum of 0 and an increment of 1 can be used to restrict numerical inputs to non-negative integers, for example.
  * `default` — a number or string giving the default value of an optional input.
* `outputs` — an array of dictionaries describing the opcode’s outputs, in order. The dictionaries can contain
  * `name` — a title-case string describing the output, suitable for presentation in a user-interface.
  * `units` — a string giving the units of the output, suitable for presentation in a user-interface.
* `syntax` — an array of dictionaries describing how the opcode can be called. This array is intended to answer the question, “What input types can I pass to this opcode, and what output types will I get as a result?” The dictionaries in this array can contain
  * `inputTypes` — a type array describing the opcode’s inputs, except that the last item can be a dictionary containing
    * `repeat` — a type array.
    * `max` — the maximum number of times the items in the `repeat` type array can be repeated.
  * `outputTypes` — a type array describing the output types for the corresponding input types. This is an array with the same constraints as the `inputTypes`.
* `requiresParentheses` — `true` if the opcode must be called with parentheses surrounding its inputs, like [`cpspch`](http://www.csounds.com/manual/html/cpspch.html).
* `author` — the opcode’s author’s name as a string.
* `dependencies` — A dictionary of key-value pairs of dependencies.

The `syntax` array of dictionaries is probably the trickiest part. In the `oscili` example, the first syntax dictionary is

```json
{ "inputTypes": [
    ["a", "k"],
    ["a", "k"],
    ["i", "i[]"],
    "i"
  ],
  "outputTypes": [
    "a"
  ]
}
```

This is saying that if you want an a‑rate output, the first two arguments to `oscili` can be either a‑ or k‑rate. The second syntax dictionary is

```json
{ "inputTypes": [
    "k",
    "k",
    ["i", "i[]"],
    "i"
  ],
  "outputTypes": [
    "k"
  ]
}
```

This is saying that if you want a k‑rate output, the first two arguments to `oscili` can only be k‑rate.

Here’s a JSON dictionary for [`outch`](http://www.csounds.com/manual/html/outch.html), using a repeating sequence of k‑ and a‑rate arguments for its inputs.

```json
{ "name": "outch",
  "description": "Output multichannel audio to specific channels",
  "descriptionURL": "http://www.csounds.com/manual/html/outch.html",
  "inputs": [
    { "name": "Channel",
      "min": 1,
      "increment": 1
    },
    { "name": "Signal" }
  ],
  "syntax": [
    { "inputTypes": [
        { "repeat": [
            "k",
            "a"
          ]
        }
      ]
    }
  ],
  "author": "John ffitch",
  "dependencies": {
    "Csound": ">=4.7.0"
  }
}
```

Here’s a JSON dictionary for [`cpspch`](http://www.csounds.com/manual/html/cpspch.html).

```json
{ "name": "cpspch",
  "description": "Convert octave and pitch-class to frequency (Hz)",
  "descriptionURL": "http://www.csounds.com/manual/html/cpspch.html",
  "inputs": [
    { "name": "Octave and Pitch Class" }
  ],
  "outputs": [
    { "name": "Frequency",
      "units": "Hz"
    }
  ],
  "syntax": [
    { "inputTypes": [
        "i"
      ],
      "outputTypes": [
        "i"
      ]
    },
    { "inputTypes": [
        "k"
      ],
      "outputTypes": [
        "k"
      ]
    }
  ],
  "requiresParentheses": true
}
```

For what it’s worth, here’s my take on this format. Besides the advantage of having a self-contained description of opcodes, JSON has a lot of language support—http://www.json.org lists a few—so it can potentially be used by many applications. Using more descriptive argument names may lead to better documentation, and title-case argument names can easily be converted to camelCase names for code examples.

The downside, of course, is that it’s not at all clear how to apply any of this to Csound’s current documentation. Are there any forks of Csound or the Csound manual with work along these lines that I might be able to contribute to? Should I fork (the manual) and have a go at this?

**Comment by jpffitch** [comment]
I thought the defaut for oscil* third arguent was -1 wic implies internal 
sine?

**Comment by nwhetsell** [comment]
Oops, that’s right. I fixed this in the example and also added an array argument for the sake of completeness.

**Comment by mantaraya36** [comment]
I presume you already know of:

https://github.com/csound/manual/blob/master/opcodes.xml

Not as rich as you propose, but useful for most front-end needs. This is
used by CsoundQt and Cabbage.

Cheers,
Andrés

**Comment by kunstmusik** [comment]
There is a python script in the manual repo that scrapes and assembles an
XML file. It could be modified to output JSON I think. Might make for a
good start.

**Comment by mantaraya36** [comment]
But it would still be hard to get all the rich information proposed from
the manual, as there are some things that are not standardized so will be
hard to parse. The solution could be to standardize to make the manual a
richer source of metadata.

**Comment by nwhetsell** [comment]
@mantaraya36 Yep, I’m aware of https://github.com/csound/manual/blob/master/opcodes.xml. Unfortunately, that file wasn’t well-suited to autocompletion in Atom.

@kunstmusik Thanks; I believe the Python script is https://github.com/csound/manual/blob/master/opcodeparser.py. I’m using https://github.com/nwhetsell/language-csound/blob/master/resources/make-opcode-completions-json.js to parse XML files in the Csound manual and combine them with input/output types from `csoundNewOpcodeList`. The JSON the script [outputs](https://github.com/nwhetsell/language-csound/blob/master/resources/opcode-completions.json) is specific to autocompletion in Atom, but I can certainly make it output something more like the above; I’m sort of generating those JSON objects on the way to creating Atom autocompletions.
